server/src/main/java/org/elasticsearch/index/mapper/DynamicFieldsBuilder.java

```diff
@@ -194,6 +194,7 @@ private static boolean applyMatchingTemplate(ParseContext context,
         Map<String, Object> mapping = dynamicTemplate.mappingForName(name, dynamicType);
         if (dynamicTemplate.isRuntimeMapping()) {
             Mapper.TypeParser.ParserContext parserContext = context.parserContext(dateFormatter);
+            parserContext = parserContext.createDynamicTemplateFieldContext(parserContext);
```

Comment:
oh boy this is super sneaky. I wonder how we can prevent this from happening again in the long run.

Reply:
Actually I think `ParseContext.parserContext()` is only ever called in the context of dynamic fields, so we can probably move this directly into that method.  Will give that a go!

Reply:
I found the createDynamicTemplateFieldContext more misleading, because it returns again a parser context and it's' very easy to miss.

Reply:
I've pushed a change which makes it unnecessary - let me know what you think